File: src/main/scala/com/seu/geomesa/spark/JtsDemo.scala
package com.seu.geomesa.spark

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions.{col, expr, lit}
import org.apache.spark.sql.types._
import org.locationtech.geomesa.spark.jts._
import org.locationtech.jts.geom.{Coordinate, GeometryFactory}

object JtsDemo {
  def main(args: Array[String]): Unit = {
    val spark = SparkSession.builder().appName("JtsDemo").master("local[*]").getOrCreate().withJTS


    val schema = StructType(Array(
      StructField("name", StringType, nullable = false),
      StructField("pointText", StringType, nullable = false),
      StructField("polygonText", StringType, nullable = false),
      StructField("latitude", DoubleType, nullable = false),
      StructField("longitude", DoubleType, nullable = false)))

    val dataFile = this.getClass.getClassLoader.getResource("data/jts-example.csv").getPath
    val df = spark.read
      .schema(schema)
      .option("sep", "-")
      .option("timestampFormat", "yyyy/MM/dd HH:mm:ss ZZ")
      .csv(dataFile)
    df.show()

    df.createOrReplaceTempView("df")
    val altered_df = spark.sql("SELECT *, st_polygonFromText(polygonText) as polygon, st_makePoint(longitude, latitude) as point from df")
    altered_df.show()

    val polygon = spark.sql("SELECT st_polygonFromText('POLYGON((-71.1776585052917 42.3902909739571,-71.1776820268866 42.3903701743239,-71.1776063012595 42.3903825660754,-71.1775826583081 42.3903033653531,-71.1776585052917 42.3902909739571))') as polygon")
    polygon.show()

    val point = new GeometryFactory().createPoint(new Coordinate(3.4, 5.6))
    println(point)

    df.where(st_contains(st_makeBBOX(0.0, 0.0, 90.0, 90.0), expr("st_makePoint(longitude, latitude)")))
      .show()
    altered_df.where(st_contains(st_makeBBOX(0.0, 0.0, 90.0, 90.0), col("point")))
      .show()
  }
}

File: src/main/java/com/seu/geomesa/spark/jts/JtsDemo.java
package com.seu.geomesa.spark.jts;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.types.*;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;
import org.locationtech.geomesa.spark.jts.package$;


import java.util.ArrayList;
import java.util.List;

public class JtsDemo {
    public static void main(String[] args) {
        SparkSession spark = SparkSession.builder().appName("JtsDemo").master("local[*]").getOrCreate();
        // 使用geomesa spark jts, scala中使用的隐式类，java中调用如下
        package$.MODULE$.initJTS(spark.sqlContext());

        String dataFile = JtsDemo.class.getResource("/data/jts-example.csv").getPath();
        System.out.println("dataFile: " + dataFile);

        // 创建schema
        List<StructField> schemaFields = new ArrayList<>();
        schemaFields.add(DataTypes.createStructField("name", DataTypes.StringType, true));
        schemaFields.add(DataTypes.createStructField("pointText", DataTypes.StringType, true));
        schemaFields.add(DataTypes.createStructField("polygonText", DataTypes.StringType, true));
        schemaFields.add(DataTypes.createStructField("latitude", DataTypes.StringType, true));
        schemaFields.add(DataTypes.createStructField("longitude", DataTypes.StringType, true));
        StructType schema = DataTypes.createStructType(schemaFields);

        // 读取数据
        Dataset<Row> df = spark.read()
                .schema(schema)
                .option("sep", "-")
                .option("timestampFormat", "yyyy/MM/dd HH:mm:ss ZZ")
                .csv(dataFile);
        df.show();

        df.createOrReplaceTempView("df");
        Dataset<Row> altered_df = spark.sql("SELECT *, st_polygonFromText(polygonText) as polygon, st_makePoint(longitude, latitude) as point from df");
        altered_df.show();
    }
}
